ffi/dart/tool/build_libs.dart

```diff
@@ -2,120 +2,237 @@
+    icu4xPath = icu4xPath.parent;
+    if (icu4xPath.parent == icu4xPath) {
+      throw ArgumentError(
+        'Running in the wrong directory, as no Cargo.toml exists above ${Platform.script}',
+      );
+    }
+  }
 
-  final isNoStd = [
-    Target.androidRiscv64,
-    Target.linuxRiscv32,
-  ].contains(target);
+  await buildLib(
+    targetOS,
+    targetArchitecture,
+    buildStatic,
+    simulator,
+    out,
+    out.pathSegments.last,
+    icu4xPath.path,
+    cargoFeatures: cargoFeatures,
+  );
+}
 
-  final rustupArguments = ['target', 'add', rustTarget];
-  final rustup = await Process.start('rustup', rustupArguments);
-  stdout.addStream(rustup.stdout);
-  stderr.addStream(rustup.stderr);
+Future<Uri> buildLibraryFromInput(
+  BuildInput input,
+  String workingDirectory,
+) async {
+  final crateNameFixed = crateName.replaceAll('-', '_');
+  final libFileName = input.config.filename(crateNameFixed);
+  final libFileUri = input.outputDirectory.resolve(libFileName);
+  await buildLib(
+    input.config.code.targetOS,
+    input.config.code.targetArchitecture,
+    input.config.buildStatic,
+    input.config.code.targetOS == OS.iOS &&
+        input.config.code.iOS.targetSdk == IOSSdk.iPhoneSimulator,
+    libFileUri,
+    libFileName,
+    workingDirectory,
+  );
+  return libFileUri;
+}
 
-  final rustupExitCode = await rustup.exitCode;
-  if (rustupExitCode != 0) {
-    throw ProcessException(
+// Copied from Dart's package:intl4x build.dart, see
+// https://github.com/dart-lang/i18n/blob/main/pkgs/intl4x/hook/build.dart
+Future<void> buildLib(
+  OS targetOS,
+  Architecture targetArchitecture,
+  bool buildStatic,
+  bool isSimulator,
+  Uri libFileUri,
+  String libFileName,
+  String workingDirectory, {
+  List<String> cargoFeatures = const [],
+}) async {
+  final isNoStd = _isNoStdTarget((targetOS, targetArchitecture));
+  final target = _asRustTarget(targetOS, targetArchitecture, isSimulator);
+  if (!isNoStd) {
+    final rustArguments = ['target', 'add', target];
+    await runProcess(
       'rustup',
-      rustupArguments,
-      'Exited with a non-zero exit code',
-      rustupExitCode,
+      rustArguments,
+      workingDirectory: workingDirectory,
     );
   }
-
-  final features = {
+  final tempDir = await Directory.systemTemp.createTemp();
+  final stdFeatures = [
+    'default_components',
+    'icu_collator,icu_datetime,icu_list,icu_decimal,icu_plurals',
+    'compiled_data',
     'buffer_provider',
-    if (isNoStd) ...['libc-alloc', 'panic-handler'],
-    if (!isNoStd) ...['logging', 'simple_logger'],
-    ...cargoFeatures,
-  };
+    'logging',
+    'simple_logger',
+    'experimental_components',
+  ];
+  final noStdFeatures = [
+    'default_components',
+    'icu_collator,icu_datetime,icu_list,icu_decimal,icu_plurals',
+    'compiled_data',
+    'buffer_provider',
+    'libc_alloc',
+    'panic_handler',
+    'experimental_components',
+  ];
+  final linkModeType = buildStatic ? 'staticlib' : 'cdylib';
```

Comment:
please try to reduce the diff here. outlining this does not improve readability imho